rrocker-lib/protos/api.proto

```diff
@@ -0,0 +1,90 @@
+syntax = "proto3";
+package rrocker.api;
+
+import "google/protobuf/empty.proto";
+/// A handle for our task, contains an UUIDv4
+message TaskHandle {
+    string uuid = 1;
+}
+
+/// Memory and CPU constraints for a task. Best effort constraints so 
+/// values above what the host can offer will mean *no* constraints
+message ResourceConstraints {
+    int32 max_cpu = 1; //CPU % of all cores on daemon host (1-100)
+    int32 max_mem_bytes = 2;  //memory in bytes
+}
+
+/// A message encoding the start task request.
+/// `cmd` is required while `args` and `constraints` are optional
+message StartTaskRequest {
+    /// cmd should be an absolute path
+    string cmd = 1; 
+    repeated string args = 2;
+    ResourceConstraints constraints = 3;
+}
+
+/// Task start reply containing a task handle
+message StartTaskReply {
+    TaskHandle handle = 1;
+}
+
+/// Designates which pipe the program output came from
+enum OutputStream {
+    Stdin = 0; //never used but let's be consistent with the linux api
+    Stdout = 1;
+    Stderr = 2;
+}
+
+/// Task output with a line of output plus which pipe it came from
+message TaskOutput {
+    string line = 1;
+    OutputStream stream = 2;
+}
+
+/// Enum describing the status of a task
+enum TaskStatus {
+    TASK_COMPLETED = 0;
+    TASK_RUNNING = 1;
+    TASK_KILLED = 2;    
+}
+
+/// The task's state is encoded as a status and an exit code if set by the task
+message TaskState {
+    TaskStatus status = 1;
+    int32 code = 2;
+}
+
+/// Reply message of the task query command containing the state
+message QueryTaskReply {
+    TaskState state = 1;
+}
+
+/// The message streamed when requesting a task's output.
+/// Each message will either be some output or a state indicating termination.
+message TaskOutputReply {
+    oneof result {
+        TaskOutput output = 1;
+        TaskState state = 2;
+    }
+}

```

Comment:
Why do you return output or state here? You have another endpoint for state right?

Reply:
Yes there's another endpoint solely for state.
The state here will only be sent as the last message to indicate why a stream ended rather than requiring an extra round trip to query the task state.

Reply:
Does the Rust GRPC library not return something like `io.EOF` that the Go library does to indicate the stream ended?

Reply:
The EOF part is handled automatically when the write handle of the gRPC stream gets dropped so the idea here is *not* to mark the end of the stream it's merely a convenience thing.
But I suppose there's a good argument to be made that it introduces extra handling on the client side for every log message received so I'll remove it.